src/main/scala/is/hail/asm4s/package.scala

```diff
@@ -25,99 +23,7 @@ package object asm4s {
-    val returnOp = ti.returnOp
-    override val slots = ti.slots
-
-    def newArray() = ti.newArray
-
-    def castFromGeneric = (_x: Code[_]) => {
-      val x = _x.asInstanceOf[Code[AnyRef]]
-      ti match {
-        case _: IntInfo.type =>
-          Code.intValue(Code.checkcast[java.lang.Integer](x)).asInstanceOf[Code[T]]
-        case _: LongInfo.type =>
-          Code.longValue(Code.checkcast[java.lang.Long](x)).asInstanceOf[Code[T]]
-        case _: FloatInfo.type =>
-          Code.floatValue(Code.checkcast[java.lang.Float](x)).asInstanceOf[Code[T]]
-        case _: DoubleInfo.type =>
-          Code.doubleValue(Code.checkcast[java.lang.Double](x)).asInstanceOf[Code[T]]
-        case _: ShortInfo.type =>
-          Code.checkcast[java.lang.Short](x).invoke[Short]("shortValue").asInstanceOf[Code[T]]
-        case _: ByteInfo.type =>
-          Code.checkcast[java.lang.Byte](x).invoke[Byte]("byteValue").asInstanceOf[Code[T]]
-        case _: BooleanInfo.type =>
-          Code.checkcast[java.lang.Boolean](x).invoke[Boolean]("booleanValue").asInstanceOf[Code[T]]
-        case _: CharInfo.type =>
-          Code.checkcast[java.lang.Character](x).invoke[Char]("charValue").asInstanceOf[Code[T]]
-        case cti: ClassInfo[_] =>
-          Code.checkcast[T](x)(cti.cct.asInstanceOf[ClassTag[T]])
-        case ati: ArrayInfo[_] =>
-          Code.checkcast[T](x)(ati.tct.asInstanceOf[ClassTag[T]])
-      }
-    }
-
-    def castToGeneric = (x: Code[T]) => ti match {
-      case _: IntInfo.type =>
-        Code.boxInt(x.asInstanceOf[Code[Int]])
-      case _: LongInfo.type =>
-        Code.boxLong(x.asInstanceOf[Code[Long]])
-      case _: FloatInfo.type =>
-        Code.boxFloat(x.asInstanceOf[Code[Float]])
-      case _: DoubleInfo.type =>
-        Code.boxDouble(x.asInstanceOf[Code[Double]])
-      case _: ShortInfo.type =>
-        Code.newInstance[java.lang.Short, Short](x.asInstanceOf[Code[Short]])
-      case _: ByteInfo.type =>
-        Code.newInstance[java.lang.Byte, Byte](x.asInstanceOf[Code[Byte]])
-      case _: BooleanInfo.type =>
-        Code.boxBoolean(x.asInstanceOf[Code[Boolean]])
-      case _: CharInfo.type =>
-        Code.newInstance[java.lang.Character, Char](x.asInstanceOf[Code[Char]])
-      case cti: ClassInfo[_] =>
-        x
-      case ati: ArrayInfo[_] =>
-        x
-    }
-
-    val generic = classInfo[java.lang.Object]
-    val isGeneric = true
-
-  }
-
-  final case class NotGenericTypeInfo[T](implicit ti: TypeInfo[T]) extends MaybeGenericTypeInfo[T] {
-    val name = ti.name
-    val loadOp = ti.loadOp
-    val storeOp = ti.storeOp
-    val aloadOp = ti.aloadOp
-    val astoreOp = ti.astoreOp
-    val returnOp = ti.returnOp
-    override val slots = ti.slots
-
-    def newArray() = ti.newArray
-
-    def castFromGeneric = (x: Code[_]) => x.asInstanceOf[Code[T]]
-    def castToGeneric = (x: Code[_]) => x
-
-    val generic = ti
-    val isGeneric = false
-
+    def classTag: ClassTag[T] = reflect.classTag
```

Comment:
I think this, the ClassTag requirement on T, the TypeInfo-only constructor for arrayInfo and ArrayInfo constructor that calls wrap are all unnecessary.  If you delete them all this version builds fine.

I guess I don't understand how this change and how it is used downstream.  Shouldn't the `ClassTag` requirement on `T` be propagated backwards to the user whomever is trying to create an `ArrayInfo`?

Reply:
Yeah you're right this was from an old version of the compiler. I removed it in the latest commit, but left the reworking of `GenericTypeInfo` because I think that's a net improvement.